### ISSUE TITLE
Performance problem: exponential running time

### ISSUE THREAD

**Issue by dr0i** [metafacture/metafacture-fix#258]
Since a month or so we experience an exponential running time, regardless using threads or not. Here a table from the logs:

|timestamp| time compared to predecessor| every 100k docs|
| ---      | ---       | ---- |
|2022-09-12 17:30:44| +30m | 100000|
|2022-09-12 18:31:59| +60m| 200000|
|2022-09-12 20:03:10| +90m| 300000|
|2022-09-12 22:05:05| +120m|  400000|
|2022-09-13 00:39:17| +150m| 500000|
|2022-09-13 03:45:12| +180m| 600000|
|2022-09-13 07:16:26| +210m| 700000|
|2022-09-13 11:20:18| +240m| 800000|

=> every 100kth docs the time needed increased by 30m in comparison to the 100kth predecessor.
One speculation regarding the cause:
some data structures are `static` ( particularly `Trie` and `HashMap` as cache) which might not being removed by the Garbage Collector (GC) and thus grow with time. Especially operations on the `Trie` become slower with its growth. Solution would be to guarantee proper GC of `Trie`.

See useful hints for analyzing at https://github.com/metafacture/metafacture-fix/issues/207.

**Comment by blackwinter** [comment]
Can you specify the exact code versions/commits you used? (lobid-resources, metafacture-fix, metafacture-core)

Was this process indexing into Elasticsearch or writing to a file?

Can you provide instructions on how to reproduce your setup?

**Comment by blackwinter** [comment]
> Since a month or so we experience an exponential running time

Does this mean your previous processes didn't exhibit this pattern.

**Comment by dr0i** [comment]
> Can you specify the exact code versions/commits you used? (lobid-resources, metafacture-fix, metafacture-core)

The aforementioned runtime data was taken from branches solving the thread problems (see #255).
Fair enough, to answer your question I set back everything, i.e. `lobid-resources` to master at commit d8af6b7f2c4434c4cc69565737c109ffa182b0cd , which used `metafacture-fix` master (`0.4.0-SNAPSHOT` at commit 131dd03edc2b2b710e6f379689d456375d8bd090). Both libraries use `metafacture-core 5.3.2`. 
Behaviour is the same as the aforementioned data:

|timestamp| time compared to predecessor| every 100k|
| --- | --- | ---|
|2022-09-13 11:41:13| - | 0 |
|2022-09-13 12:12:38| +30m | 100k|
|2022-09-13 13:11:14| +60m | 200k|
|2022-09-13 14:42:48 | +90m | 300k|

> Was this process indexing into Elasticsearch or writing to a file?

Into ES. I am bold to say: writing to file or /dev/null wouldn't make a  difference.

> Can you provide instructions on how to reproduce your setup?

```
git clone https://github.com/hbz/lobid-resources.git;
cd lobid-resources/src/test/resources/;
bash buildAndETLTestAlmaFix.sh;  #configure "baseline.xml.bgzf" as input data;
```
(make sure that you have installed metafix to local maven repo in the proper version).
Bash script would index into our ES-Cluster, _if_ you are in the proper internal network zone (I assume you are not), but even if you are in that network,  you just may do it (no harm).

Even tried out not using `ObjectDecoupler` at all, i.e. rule out any possible threading problem, like with:
``` Metafix metafix = new Metafix(fixFileName, fixVariables);
      LOG.info("Setting strictness to EXPRESSION => metafix will not break if a field makes trouble but rather logs warning");
      metafix.setStrictness(Metafix.Strictness.EXPRESSION);
      metafix.setStrictnessHandlesProcessExceptions(true);

      opener.setReceiver(new XmlDecoder())//
          .setReceiver(marcXmlHandler)//
          .setReceiver(metafix)
          .setReceiver(batchLogger)//
          .setReceiver(jsonEncoder)//
          .setReceiver(etikettJson)//
          .setReceiver(new JsonToElasticsearchBulkMap(keyToGetMainId, "resource",
              "ignored"))//
          .setReceiver(getElasticsearchIndexer());
          opener.process(inputPath);
          opener.closeStream();
```
(substituting the pipeline starting at https://github.com/hbz/lobid-resources/blob/master/src/main/java/org/lobid/resources/run/AlmaMarcXmlFix2lobidJsonEs.java#L132 ) You could also comment out any ` .addReceiver(receiverThread())` except one.

> Does this mean your previous processes didn't exhibit this pattern?

Yep. First mention of runtime decline was at https://github.com/hbz/lobid-resources/pull/1416#issuecomment-1228486903.

Going to try these ideas:
- [x] rule out race conditions as possible memory leaks etc. (see #255)
- [ ] trying out https://github.com/hbz/lobid-resources/pull/1426 provided by @TobiasNx who identified one major difference in your limetrans and our fixes.
- [ ] get rid of some `static` declarations in metafix (as proposed here in the first comment)
- [ ] trying to profile
- [ ] going back in commit history to pin down the problematic commit

**Comment by blackwinter** [comment]
Thanks a lot! I'll try to reproduce the behaviour and report back.

**Comment by blackwinter** [comment]
So I was able to reproduce your findings. I've indexed into a _local_ Elasticsearch instance (version 2.2[^1]) with 6 threads on a quad-core machine (logging every 10k records, cut off after 600k records per thread).

Versions:

| Module           | Version/Commit              |
|------------------|-----------------------------|
| lobid-resources  | hbz/lobid-resources@d8af6b7 |
| metafacture-fix  | 0.4.0-SNAPSHOT (2d98aae)    |
| metafacture-core | 5.3.2                       |

Command:

```sh
mvn exec:java\
  -Dexec.mainClass=org.lobid.resources.run.AlmaMarcXmlFix2lobidJsonEs\
  -Dexec.cleanupDaemonThreads=false -DjsonLdContext=http://test.lobid.org/resources/context.jsonld\
  -Dexec.args="/path/to/export/alma/prod/baseline.xml.bgzf resources-almatest NOALIAS localhost elasticsearch create"
```

Here are the results (averaged across all threads):

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 100000 | 25m27s   |            |            |
| 200000 | 49m3s    | +1416.50   | +92.77%    |
| 300000 | 1h22m39s | +2015.67   | +68.48%    |
| 400000 | 1h38m1s  | +921.83    | +18.59%    |
| 500000 | 2h0m22s  | +1340.83   | +22.80%    |
| 600000 | 2h24m51s | +1469.00   | +20.34%    |

<details><summary>Details (10k batches)</summary>

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 1m41s    |            |            |
| 20000  | 1m42s    | +1.33      | +1.32%     |
| 30000  | 1m37s    | -5.33      | -5.22%     |
| 40000  | 1m53s    | +16.00     | +16.52%    |
| 50000  | 2m15s    | +22.00     | +19.50%    |
| 60000  | 2m30s    | +15.00     | +11.12%    |
| 70000  | 2m43s    | +13.17     | +8.79%     |
| 80000  | 3m18s    | +35.33     | +21.68%    |
| 90000  | 4m1s     | +42.50     | +21.43%    |
| 100000 | 3m47s    | -13.50     | -5.61%     |
| 110000 | 3m50s    | +3.00      | +1.32%     |
| 120000 | 4m5s     | +14.83     | +6.44%     |
| 130000 | 4m31s    | +25.67     | +10.47%    |
| 140000 | 4m48s    | +16.83     | +6.22%     |
| 150000 | 4m56s    | +7.83      | +2.72%     |
| 160000 | 4m44s    | -11.17     | -3.78%     |
| 170000 | 4m56s    | +11.17     | +3.93%     |
| 180000 | 5m12s    | +16.17     | +5.47%     |
| 190000 | 5m38s    | +26.67     | +8.56%     |
| 200000 | 6m24s    | +45.67     | +13.50%    |
| 210000 | 6m29s    | +5.00      | +1.30%     |
| 220000 | 6m31s    | +2.33      | +0.60%     |
| 230000 | 7m56s    | +84.67     | +21.64%    |
| 240000 | 8m28s    | +32.00     | +6.72%     |
| 250000 | 8m34s    | +5.83      | +1.15%     |
| 260000 | 7m53s    | -41.17     | -8.01%     |
| 270000 | 8m21s    | +28.33     | +5.99%     |
| 280000 | 9m6s     | +45.17     | +9.02%     |
| 290000 | 9m33s    | +26.50     | +4.85%     |
| 300000 | 9m48s    | +15.67     | +2.74%     |
| 310000 | 9m41s    | -7.33      | -1.25%     |
| 320000 | 9m18s    | -23.17     | -3.99%     |
| 330000 | 9m17s    | -0.83      | -0.15%     |
| 340000 | 9m10s    | -7.33      | -1.32%     |
| 350000 | 9m37s    | +27.17     | +4.94%     |
| 360000 | 9m39s    | +2.33      | +0.40%     |
| 370000 | 10m6s    | +26.83     | +4.63%     |
| 380000 | 10m10s   | +4.17      | +0.69%     |
| 390000 | 10m22s   | +11.83     | +1.94%     |
| 400000 | 10m41s   | +19.17     | +3.08%     |
| 410000 | 10m53s   | +11.50     | +1.79%     |
| 420000 | 11m8s    | +14.83     | +2.27%     |
| 430000 | 11m25s   | +17.50     | +2.62%     |
| 440000 | 11m39s   | +13.67     | +2.00%     |
| 450000 | 11m52s   | +13.17     | +1.88%     |
| 460000 | 12m27s   | +35.50     | +4.99%     |
| 470000 | 12m19s   | -8.50      | -1.14%     |
| 480000 | 12m45s   | +26.33     | +3.56%     |
| 490000 | 12m50s   | +4.67      | +0.61%     |
| 500000 | 13m5s    | +15.00     | +1.95%     |
| 510000 | 13m20s   | +15.33     | +1.95%     |
| 520000 | 13m34s   | +14.17     | +1.77%     |
| 530000 | 13m53s   | +18.83     | +2.31%     |
| 540000 | 14m2s    | +8.83      | +1.06%     |
| 550000 | 14m33s   | +31.00     | +3.68%     |
| 560000 | 14m32s   | -0.83      | -0.10%     |
| 570000 | 14m52s   | +19.67     | +2.25%     |
| 580000 | 15m4s    | +12.67     | +1.42%     |
| 590000 | 15m17s   | +12.67     | +1.40%     |
| 600000 | 15m42s   | +25.17     | +2.74%     |
</details>

And with a single thread (no `ObjectThreader`, otherwise identical setup):

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 100000 | 14m28s   |            |            |
| 200000 | 26m38s   | +730.00    | +84.10%    |
| 300000 | 42m5s    | +927.00    | +58.01%    |
| 400000 | 56m24s   | +859.00    | +34.02%    |
| 500000 | 1h11m47s | +923.00    | +27.28%    |
| 600000 | 1h26m54s | +907.00    | +21.06%    |

<details><summary>Details (10k batches)</summary>

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 48s      |            |            |
| 20000  | 57s      | +9.00      | +18.75%    |
| 30000  | 1m3s     | +6.00      | +10.53%    |
| 40000  | 1m13s    | +10.00     | +15.87%    |
| 50000  | 1m19s    | +6.00      | +8.22%     |
| 60000  | 1m30s    | +11.00     | +13.92%    |
| 70000  | 1m41s    | +11.00     | +12.22%    |
| 80000  | 1m46s    | +5.00      | +4.95%     |
| 90000  | 2m7s     | +21.00     | +19.81%    |
| 100000 | 2m4s     | -3.00      | -2.36%     |
| 110000 | 1m56s    | -8.00      | -6.45%     |
| 120000 | 2m16s    | +20.00     | +17.24%    |
| 130000 | 2m21s    | +5.00      | +3.68%     |
| 140000 | 2m26s    | +5.00      | +3.55%     |
| 150000 | 2m38s    | +12.00     | +8.22%     |
| 160000 | 2m42s    | +4.00      | +2.53%     |
| 170000 | 2m52s    | +10.00     | +6.17%     |
| 180000 | 2m59s    | +7.00      | +4.07%     |
| 190000 | 3m9s     | +10.00     | +5.59%     |
| 200000 | 3m19s    | +10.00     | +5.29%     |
| 210000 | 3m26s    | +7.00      | +3.52%     |
| 220000 | 3m38s    | +12.00     | +5.83%     |
| 230000 | 3m50s    | +12.00     | +5.50%     |
| 240000 | 3m59s    | +9.00      | +3.91%     |
| 250000 | 4m9s     | +10.00     | +4.18%     |
| 260000 | 4m14s    | +5.00      | +2.01%     |
| 270000 | 4m27s    | +13.00     | +5.12%     |
| 280000 | 4m36s    | +9.00      | +3.37%     |
| 290000 | 4m42s    | +6.00      | +2.17%     |
| 300000 | 5m4s     | +22.00     | +7.80%     |
| 310000 | 5m3s     | -1.00      | -0.33%     |
| 320000 | 5m8s     | +5.00      | +1.65%     |
| 330000 | 5m21s    | +13.00     | +4.22%     |
| 340000 | 5m27s    | +6.00      | +1.87%     |
| 350000 | 5m41s    | +14.00     | +4.28%     |
| 360000 | 5m48s    | +7.00      | +2.05%     |
| 370000 | 5m55s    | +7.00      | +2.01%     |
| 380000 | 5m49s    | -6.00      | -1.69%     |
| 390000 | 5m55s    | +6.00      | +1.72%     |
| 400000 | 6m17s    | +22.00     | +6.20%     |
| 410000 | 6m17s    | +0.00      | +0.00%     |
| 420000 | 6m25s    | +8.00      | +2.12%     |
| 430000 | 6m32s    | +7.00      | +1.82%     |
| 440000 | 6m42s    | +10.00     | +2.55%     |
| 450000 | 6m53s    | +11.00     | +2.74%     |
| 460000 | 7m0s     | +7.00      | +1.69%     |
| 470000 | 8m20s    | +80.00     | +19.05%    |
| 480000 | 7m43s    | -37.00     | -7.40%     |
| 490000 | 7m53s    | +10.00     | +2.16%     |
| 500000 | 8m2s     | +9.00      | +1.90%     |
| 510000 | 8m12s    | +10.00     | +2.07%     |
| 520000 | 8m26s    | +14.00     | +2.85%     |
| 530000 | 8m36s    | +10.00     | +1.98%     |
| 540000 | 8m41s    | +5.00      | +0.97%     |
| 550000 | 8m44s    | +3.00      | +0.58%     |
| 560000 | 8m56s    | +12.00     | +2.29%     |
| 570000 | 8m41s    | -15.00     | -2.80%     |
| 580000 | 8m48s    | +7.00      | +1.34%     |
| 590000 | 8m53s    | +5.00      | +0.95%     |
| 600000 | 8m57s    | +4.00      | +0.75%     |
</details>

[^1]: I had to make local changes in order to establish compatibility.

**Comment by blackwinter** [comment]
> I am bold to say: writing to file or /dev/null wouldn't make a difference.

That was indeed a bold call! But I've confirmed it for you ;)

Same as before, except writing to a file instead of indexing into Elasticsearch (single thread, replaced `ElasticsearchIndexer` with `ObjectWriter`, cut off after 400k records):

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 100000 | 13m46s   |            |            |
| 200000 | 25m54s   | +728.00    | +88.14%    |
| 300000 | 41m29s   | +935.00    | +60.17%    |
| 400000 | 56m2s    | +873.00    | +35.07%    |

<details><summary>Details (10k batches)</summary>

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 48s      |            |            |
| 20000  | 53s      | +5.00      | +10.42%    |
| 30000  | 59s      | +6.00      | +11.32%    |
| 40000  | 1m10s    | +11.00     | +18.64%    |
| 50000  | 1m15s    | +5.00      | +7.14%     |
| 60000  | 1m26s    | +11.00     | +14.67%    |
| 70000  | 1m37s    | +11.00     | +12.79%    |
| 80000  | 1m43s    | +6.00      | +6.19%     |
| 90000  | 1m55s    | +12.00     | +11.65%    |
| 100000 | 2m0s     | +5.00      | +4.35%     |
| 110000 | 1m54s    | -6.00      | -5.00%     |
| 120000 | 2m12s    | +18.00     | +15.79%    |
| 130000 | 2m16s    | +4.00      | +3.03%     |
| 140000 | 2m22s    | +6.00      | +4.41%     |
| 150000 | 2m32s    | +10.00     | +7.04%     |
| 160000 | 2m38s    | +6.00      | +3.95%     |
| 170000 | 2m48s    | +10.00     | +6.33%     |
| 180000 | 2m57s    | +9.00      | +5.36%     |
| 190000 | 3m3s     | +6.00      | +3.39%     |
| 200000 | 3m12s    | +9.00      | +4.92%     |
| 210000 | 3m23s    | +11.00     | +5.73%     |
| 220000 | 3m35s    | +12.00     | +5.91%     |
| 230000 | 3m48s    | +13.00     | +6.05%     |
| 240000 | 3m55s    | +7.00      | +3.07%     |
| 250000 | 4m6s     | +11.00     | +4.68%     |
| 260000 | 4m11s    | +5.00      | +2.03%     |
| 270000 | 4m25s    | +14.00     | +5.58%     |
| 280000 | 4m33s    | +8.00      | +3.02%     |
| 290000 | 4m39s    | +6.00      | +2.20%     |
| 300000 | 4m54s    | +15.00     | +5.38%     |
| 310000 | 5m1s     | +7.00      | +2.38%     |
| 320000 | 5m8s     | +7.00      | +2.33%     |
| 330000 | 5m19s    | +11.00     | +3.57%     |
| 340000 | 5m25s    | +6.00      | +1.88%     |
| 350000 | 5m39s    | +14.00     | +4.31%     |
| 360000 | 5m46s    | +7.00      | +2.06%     |
| 370000 | 5m52s    | +6.00      | +1.73%     |
| 380000 | 5m47s    | -5.00      | -1.42%     |
| 390000 | 5m54s    | +7.00      | +2.02%     |
| 400000 | 6m11s    | +17.00     | +4.80%     |
</details>

**Comment by blackwinter** [comment]
Just for kicks, I've also run your Fix with Limetrans. Same outcome:

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 100000 | 12m59s   |            |            |
| 200000 | 28m5s    | +906.00    | +116.30%   |
| 300000 | 45m21s   | +1036.00   | +61.48%    |
| 400000 | 1h1m55s  | +994.00    | +36.53%    |

<details><summary>Details (10k batches)</summary>

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 34s      |            |            |
| 20000  | 42s      | +8.00      | +23.53%    |
| 30000  | 52s      | +10.00     | +23.81%    |
| 40000  | 1m3s     | +11.00     | +21.15%    |
| 50000  | 1m12s    | +9.00      | +14.29%    |
| 60000  | 1m23s    | +11.00     | +15.28%    |
| 70000  | 1m34s    | +11.00     | +13.25%    |
| 80000  | 1m42s    | +8.00      | +8.51%     |
| 90000  | 1m55s    | +13.00     | +12.75%    |
| 100000 | 2m2s     | +7.00      | +6.09%     |
| 110000 | 2m2s     | +0.00      | +0.00%     |
| 120000 | 2m18s    | +16.00     | +13.11%    |
| 130000 | 2m26s    | +8.00      | +5.80%     |
| 140000 | 2m33s    | +7.00      | +4.79%     |
| 150000 | 2m45s    | +12.00     | +7.84%     |
| 160000 | 2m52s    | +7.00      | +4.24%     |
| 170000 | 3m3s     | +11.00     | +6.40%     |
| 180000 | 3m13s    | +10.00     | +5.46%     |
| 190000 | 3m21s    | +8.00      | +4.15%     |
| 200000 | 3m32s    | +11.00     | +5.47%     |
| 210000 | 3m42s    | +10.00     | +4.72%     |
| 220000 | 3m56s    | +14.00     | +6.31%     |
| 230000 | 4m7s     | +11.00     | +4.66%     |
| 240000 | 4m18s    | +11.00     | +4.45%     |
| 250000 | 4m28s    | +10.00     | +3.88%     |
| 260000 | 4m35s    | +7.00      | +2.61%     |
| 270000 | 4m49s    | +14.00     | +5.09%     |
| 280000 | 4m58s    | +9.00      | +3.11%     |
| 290000 | 5m7s     | +9.00      | +3.02%     |
| 300000 | 5m21s    | +14.00     | +4.56%     |
| 310000 | 5m30s    | +9.00      | +2.80%     |
| 320000 | 5m37s    | +7.00      | +2.12%     |
| 330000 | 5m51s    | +14.00     | +4.15%     |
| 340000 | 5m59s    | +8.00      | +2.28%     |
| 350000 | 6m13s    | +14.00     | +3.90%     |
| 360000 | 6m20s    | +7.00      | +1.88%     |
| 370000 | 6m29s    | +9.00      | +2.37%     |
| 380000 | 6m29s    | +0.00      | +0.00%     |
| 390000 | 6m36s    | +7.00      | +1.80%     |
| 400000 | 6m51s    | +15.00     | +3.79%     |
</details>

<details><summary>Config & command</summary>

```json
// src/conf/test/alma-lobid.json
{
  "isil": "DE-605",
  "alma": {},
  "input": {
    "queue": {
      "path": "/path/to/export/alma/prod/",
      "pattern": "baseline.xml.bgzf"
    }
  },
  "output": {
    "json": "lobid.json.gz"
  },
  "transformation-rules": "../lobid-resources/src/main/resources/alma/alma.fix"
}
```

```sh
./gradlew -Dhbz.limetrans.type=fix testAlmaLobid -Pversions.metafix=master -Pgit.baseUrl=..
```
</details>

And now our own Alma Fix for comparison:

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 100000 | 9m49s    |            |            |
| 200000 | 9m50s    | +1.00      | +0.17%     |
| 300000 | 9m37s    | -13.00     | -2.20%     |
| 400000 | 9m44s    | +7.00      | +1.21%     |

<details><summary>Details (10k batches)</summary>

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 1m2s     |            |            |
| 20000  | 59s      | -3.00      | -4.84%     |
| 30000  | 59s      | +0.00      | +0.00%     |
| 40000  | 59s      | +0.00      | +0.00%     |
| 50000  | 57s      | -2.00      | -3.39%     |
| 60000  | 58s      | +1.00      | +1.75%     |
| 70000  | 1m1s     | +3.00      | +5.17%     |
| 80000  | 57s      | -4.00      | -6.56%     |
| 90000  | 58s      | +1.00      | +1.75%     |
| 100000 | 59s      | +1.00      | +1.72%     |
| 110000 | 58s      | -1.00      | -1.69%     |
| 120000 | 1m2s     | +4.00      | +6.90%     |
| 130000 | 1m0s     | -2.00      | -3.23%     |
| 140000 | 59s      | -1.00      | -1.67%     |
| 150000 | 59s      | +0.00      | +0.00%     |
| 160000 | 58s      | -1.00      | -1.69%     |
| 170000 | 59s      | +1.00      | +1.72%     |
| 180000 | 59s      | +0.00      | +0.00%     |
| 190000 | 58s      | -1.00      | -1.69%     |
| 200000 | 58s      | +0.00      | +0.00%     |
| 210000 | 59s      | +1.00      | +1.72%     |
| 220000 | 57s      | -2.00      | -3.39%     |
| 230000 | 58s      | +1.00      | +1.75%     |
| 240000 | 57s      | -1.00      | -1.72%     |
| 250000 | 57s      | +0.00      | +0.00%     |
| 260000 | 58s      | +1.00      | +1.75%     |
| 270000 | 58s      | +0.00      | +0.00%     |
| 280000 | 58s      | +0.00      | +0.00%     |
| 290000 | 57s      | -1.00      | -1.72%     |
| 300000 | 58s      | +1.00      | +1.75%     |
| 310000 | 59s      | +1.00      | +1.72%     |
| 320000 | 57s      | -2.00      | -3.39%     |
| 330000 | 58s      | +1.00      | +1.75%     |
| 340000 | 58s      | +0.00      | +0.00%     |
| 350000 | 58s      | +0.00      | +0.00%     |
| 360000 | 58s      | +0.00      | +0.00%     |
| 370000 | 59s      | +1.00      | +1.72%     |
| 380000 | 59s      | +0.00      | +0.00%     |
| 390000 | 59s      | +0.00      | +0.00%     |
| 400000 | 59s      | +0.00      | +0.00%     |
</details>

**Lo and behold, it's constant!** Phew ;)

So there _must_ be something about your Fix that's causing the excessive growth.

**Comment by dr0i** [comment]
Going to use https://itnext.io/git-binary-search-de86e9aa334a as @fsteeg suggested (rather than trying each commit on its own).

**Comment by blackwinter** [comment]
> I think the best way is to identify the "bad" commit .

For starters, I commented out different combinations of `include()`s in your main Fix and found `subjects.fix` to be the culprit:

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 38s      |            |            |
| 20000  | 35s      | -3.00      | -7.89%     |
| 30000  | 33s      | -2.00      | -5.71%     |
| 40000  | 34s      | +1.00      | +3.03%     |
| 50000  | 31s      | -3.00      | -8.82%     |
| 60000  | 32s      | +1.00      | +3.23%     |
| 70000  | 35s      | +3.00      | +9.38%     |
| 80000  | 30s      | -5.00      | -14.29%    |
| 90000  | 33s      | +3.00      | +10.00%    |
| 100000 | 32s      | -1.00      | -3.03%     |

----

BTW: What's interesting, though, is that I never observed your ["triangular"](https://en.wikipedia.org/wiki/1_+_2_+_3_+_4_+_⋯) series, my progression was always more or less linear.

**Comment by dr0i** [comment]
I don't seem to comprehend your latest measurement - numbers seem to be fine - there is no exponential (triangular) runtime?

**Comment by blackwinter** [comment]
Exactly, that's without including `subjects.fix`.

**Comment by dr0i** [comment]
Great news :) (by coincidence @TobiasNx was just telling me that he suspects the `macro` ).
He will get rid of the `macro` and you will have a look at `metafix` to get rid of that bug I guess.

**Comment by blackwinter** [comment]
> He will get rid of the `macro`

Why? Just don't declare it over and over for each record, that's what `do once()` is for.

**Comment by blackwinter** [comment]
> We just understood it now that `do put_macro` is setting a global makro and not just for each transformation.

Yes, it acts just like [script-level functions](https://github.com/metafacture/metafacture-fix#script-level-functions) in this regard.

> This explains the exponential behavior, since new makro are put again and again.

No, that doesn't explain it per se. There's no apparent reason it should take longer every time it's called.

**Comment by blackwinter** [comment]
I think I found the explanation: Every time the macro is (re-)declared we add the "static" local variables (`options`) to its internal list of `vars`. So this list grows with each record and merging it (`getVars()`) takes longer and longer.

Let's see what we can do about it...

**Comment by blackwinter** [comment]
> `static` can be a beast.

This has nothing to do with Java's `static` variables. See the [documentation for `put_macro()`](https://github.com/metafacture/metafacture-fix#do-put_macro) for an explanation of the different types of variables a macro accepts.

**Comment by blackwinter** [comment]
Should be fixed with #259.

Before:

_In progress_

<details><summary>Details (10k batches)</summary>

_In progress_
</details>

After:

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 100000 | 8m10s    |            |            |
| 200000 | 8m15s    | +5.00      | +1.02%     |
| 300000 | 8m10s    | -5.00      | -1.01%     |
| 400000 | 5m11s    | -179.00    | -36.53%    |

<details><summary>Details (10k batches)</summary>

| Batch  | Duration | Abs. Diff. | Rel. Diff. |
|--------|----------|------------|------------|
| 10000  | 1m24s    |            |            |
| 20000  | 1m19s    | -5.00      | -5.95%     |
| 30000  | 1m7s     | -12.00     | -15.19%    |
| 40000  | 36s      | -31.00     | -46.27%    |
| 50000  | 38s      | +2.00      | +5.56%     |
| 60000  | 37s      | -1.00      | -2.63%     |
| 70000  | 37s      | +0.00      | +0.00%     |
| 80000  | 37s      | +0.00      | +0.00%     |
| 90000  | 37s      | +0.00      | +0.00%     |
| 100000 | 38s      | +1.00      | +2.70%     |
| 110000 | 37s      | -1.00      | -2.63%     |
| 120000 | 36s      | -1.00      | -2.70%     |
| 130000 | 36s      | +0.00      | +0.00%     |
| 140000 | 41s      | +5.00      | +13.89%    |
| 150000 | 1m19s    | +38.00     | +92.68%    |
| 160000 | 1m17s    | -2.00      | -2.53%     |
| 170000 | 47s      | -30.00     | -38.96%    |
| 180000 | 42s      | -5.00      | -10.64%    |
| 190000 | 41s      | -1.00      | -2.38%     |
| 200000 | 59s      | +18.00     | +43.90%    |
| 210000 | 1m8s     | +9.00      | +15.25%    |
| 220000 | 57s      | -11.00     | -16.18%    |
| 230000 | 1m0s     | +3.00      | +5.26%     |
| 240000 | 1m37s    | +37.00     | +61.67%    |
| 250000 | 49s      | -48.00     | -49.48%    |
| 260000 | 30s      | -19.00     | -38.78%    |
| 270000 | 30s      | +0.00      | +0.00%     |
| 280000 | 29s      | -1.00      | -3.33%     |
| 290000 | 31s      | +2.00      | +6.90%     |
| 300000 | 39s      | +8.00      | +25.81%    |
| 310000 | 32s      | -7.00      | -17.95%    |
| 320000 | 31s      | -1.00      | -3.12%     |
| 330000 | 31s      | +0.00      | +0.00%     |
| 340000 | 30s      | -1.00      | -3.23%     |
| 350000 | 31s      | +1.00      | +3.33%     |
| 360000 | 31s      | +0.00      | +0.00%     |
| 370000 | 31s      | +0.00      | +0.00%     |
| 380000 | 31s      | +0.00      | +0.00%     |
| 390000 | 31s      | +0.00      | +0.00%     |
| 400000 | 32s      | +1.00      | +3.23%     |
</details>

**Comment by dr0i** [comment]
> This has nothing to do with Java's static variables.
Oh, right. Then it seems at least that the concept of `static` can be beasty (if I understand the docu correctly).
